substrate-node/support/src/types.rs

```diff
@@ -204,16 +212,103 @@ pub struct Interface<Name, Mac, Ips> {
 }
 
 #[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Encode, Decode, Default, Debug, TypeInfo)]
-pub struct PublicConfig<IP4, IP6, Domain> {
+pub struct PublicConfig {
     pub ip4: IP4,
-    pub ip6: IP6,
-    pub domain: Domain,
+    pub ip6: Option<IP6>,
+    pub domain: Option<BoundedVec<u8, ConstU32<MAX_DOMAIN_NAME_LENGTH>>>,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Encode, Decode, Default, Debug, TypeInfo)]
+pub struct IP4 {
+    pub ip: BoundedVec<u8, ConstU32<MAX_IP4_LENGTH>>,
+    pub gw: BoundedVec<u8, ConstU32<MAX_GW4_LENGTH>>,
+}
+
+pub enum PublicIpError {
+    InvalidIp4,
+    InvalidGw4,
+    InvalidIp6,
+    InvalidGw6,
+    InvalidPublicIp,
+    InvalidDomain,
+}
+
+impl IP4 {
+    pub fn is_valid(&self) -> Result<(), PublicIpError> {
+        let gw4 = IPv4::parse(&self.gw).map_err(|_| PublicIpError::InvalidGw4)?;
+        let ip4 = IPv4Cidr::parse(&self.ip).map_err(|_| PublicIpError::InvalidIp4)?;
+
+        if gw4.is_public()
+            && gw4.is_unicast()
+            && ip4.is_public()
+            && ip4.is_unicast()
+            && ip4.contains(gw4)
+        {
+            return Ok(());
+        } else {
+            return Err(PublicIpError::InvalidPublicIp);
```

Comment:
These returns can be elided since this is the last expression so its result will implicitly be returned (`;` needs to be removed though)